compiler/rustc_lint/src/internal.rs

```diff
@@ -676,3 +676,26 @@ impl<'tcx> LateLintPass<'tcx> for SymbolInternStringLiteral {
         }
     }
 }
+
+declare_tool_lint! {
+    /// The `direct_use_of_rustc_type_ir` lint detects usage of `rustc_type_ir`.
+    ///
+    /// This module should only be used within the trait solver and some desirable
+    /// crates like rustc_middle.
+    pub rustc::DIRECT_USE_OF_RUSTC_TYPE_IR,
+    Allow,
+    "usage `rustc_type_ir` abstraction outside of trait system",
+    report_in_external_macro: true
+}
+declare_lint_pass!(UseOfTypeIr => [DIRECT_USE_OF_RUSTC_TYPE_IR]);
+
+impl<'tcx> LateLintPass<'tcx> for UseOfTypeIr {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx rustc_hir::Item<'tcx>) {
+        let rustc_hir::ItemKind::Use(path, _) = item.kind else { return };
+        if let Some(segment) = path.segments.first()
+            && segment.ident.to_string() == "rustc_type_ir"
```

Comment:
Probably should use a diagnostic item on the module, and the segment's partial res, rather than just looking for `rustc_type_ir`. Look at the existing lint for denying `rustc_type_ir::inherent`.

Reply:
Also, why make this into a new `LateLintPass` rather than fold this into the existing `TypeIr` pass?